Automattic/woocommerce-payments
# Add IPP print receipt preview

Fixes #3561 

#### Changes proposed in this Pull Request
This adds a new link to the Receipt details section in the Card Reades admin menu: Preview a printed receipt. When a user modifies any of the settings and clicks on this link, they can see a preview of the printed receipt with the modified settings. It is worth noting that there is no need to save the settings to see the preview, and generating the preview does not persist the data.

This feature adds some complexity to deal with the preview generation. It adds a new endpoint under `/payments/readers/receipts/print/preview`. This endpoint relies on the existing `get_receipt_markup` method on `WC_Payments_In_Person_Payments_Receipts_Service`. This has the benefit of not having to maintain two templates (one on the front end and another on the back end), since the method that generates the markup of the receipt is the same for printing and previewing purposes.

This new endpoint accepts `POST` requests. It will use the posted settings to generate the preview. It will default to the merchant settings if no settings are provided.

The preview markup is returned to the front end as a string. I use [dangerouslySetInnerHTML](https://reactjs.org/docs/dom-elements.html#dangerouslysetinnerhtml) to inject in into the DOM. Since we are escaping the html when the template is parsed I believe this is safe, but please let me know if there are any security concerns.

This is a screencast of the feature in action:

https://user-images.githubusercontent.com/1553182/152504083-6cfedc40-03d7-46a8-9f03-9e52e08224bc.mp4










#### Testing instructions





* Enable Card Readers in admin menu by following instructions in [this PR](https://github.com/Automattic/woocommerce-payments/pull/3410)
* Go to `Payments -> Card Readers -> Receipt details`
* Modify settings
* Click on `Preview a printed receipt` link
* Observe that the generated receipt contains the modified settings data
* Reload the page
* Observe that the generated receipt contains the originial unmodified data


-------------------


**Post merge**



- [x] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) : [Link here](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions-for-WC-Payments-3.9.0#add-ipp-print-receipt-preview-to-card-readers-page)
